pkg/status/status.go

```diff
@@ -1,65 +1,89 @@
 package status
 
 import (
-	"context"
+	"fmt"
+	"sync"
+	"time"
 
 	configv1 "github.com/openshift/api/config/v1"
 	configclient "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
 	cohelpers "github.com/openshift/library-go/pkg/config/clusteroperator/v1helpers"
 	operatorhelpers "github.com/openshift/library-go/pkg/operator/v1helpers"
 	log "github.com/sirupsen/logrus"
-	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
-	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
 )
 
-const clusterOperatorName = "marketplace-operator"
+const (
+	clusterOperatorName     = "marketplace-operator"
+	minSyncsBeforeReporting = 4
+	// At this point, a failure is critical so all syncs must be successful
+	// to report a working operator
+	successRatio           = 1.0
+	maxSyncsBeforeTruncate = 10000
+	syncTruncateValue      = 100
+)
+
+// SyncCh is used to report a sync event
+// The ratio of Successful Syncs / Total Syncs will be used to report operator health to the CVO
+var syncCh = make(chan error)
+
+// status will be a singleton
+var instance *status
+var mu sync.Mutex
 
 type status struct {
 	configClient    *configclient.ConfigV1Client
 	coAPINotPresent bool
 	namespace       string
 	clusterOperator *configv1.ClusterOperator
 	version         string
+	successfulSyncs int
+	syncs           int
```

Comment:
both `monitorClusterStatus` and `syncChannelConsumer` are running in different go routines, one is reading from these stats and the other is writing to these. In such case, we run into race condition. Ideally we should update these stats in an atomic manner.

Reply:
@tkashem `monitorClusterStatus` is intended to be async from `syncChannelConsumer`. It doesn't matter if `monitorClusterStatus` reads the wrong value, as long as they are eventually consistent (ie. 5 seconds later when the sync runs again). If the only concern is thread safety, then we should just put a lock around the writes to the stats.

Reply:
yeah, each can run concurrently if we desire to. But by splitting the read and write into their own go routines we now have to make sure to eliminate race conditions.

Reply:
I suggest that we can make these variables [atomic](https://golang.org/pkg/sync/atomic/).

Reply:
You could have interleaved read between the two atomic writes ( write `successfulSyncs` and then write  `syncs` ). These two writes constitute a single logical transaction. On the other hand, these are stats where we can tolerate some staleness. I am fine either way.